torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -202,6 +198,10 @@ void ProcessGroupAgent::join() {
       SendWork(allWorkerInfo_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
   listenerThread_.join();
+  rpcRunning_.store(false);
```

Comment:
the old `shutdown` variable is now replaced with `rpcRunning`, and the store call is moved here so that the listener thread can process the last shutdown message and shut itself down that way. Then, we need to set this variable and notify the watchdog thread for future timeouts to join as well.

Reply:
yes, the difference is that join sends messages to other workers and waits on those messages being processed, while shutdown does not wait on them being processed. Right now both of them terminate the workers, but after a discussion with @satgera @pritamdamania87 and a few others the plan was eventually to be able to call something like.

```
rpc.wait_all_workers() # doesnt shutdown the rpc agent, but synchronizes workers, similar to the old sync_rpc
rpc.shutdown() # does the shutdown
```